File: 1759-count-homo-strings.py
class Solution:
    def countHomogenous(self, s: str) -> int:
        # splicing
        def splice(s:str):
            sp = []
            if not s:
                return sp
            prevchar = s[0]
            prevstr = prevchar
            for c in s[1:]:
                if prevchar != c:
                    sp.append(prevstr)
                    prevstr = ""
                    prevchar = c
                prevstr += c
            sp.append(prevstr)
            return sp
        ans = sum(map(lambda x: (x * (x + 1))//2 , [len(s) for s in splice(s)]))
        return ans % (10**9 + 7)
        # return splice(s)
        
if __name__ == '__main__':
    ans = Solution().countHomogenous("abbcccaa")
    print(ans)

File: 1759-count-homo-strings.rs
const modulo : i64 = 10_i64.pow(9) + 7;

impl Solution {
    pub fn count_homogenous(s: String) -> i32 {
        if s.is_empty() {
            return 0;
        }
        let mut iterator = s.chars();
        let mut prevchar = iterator.next().unwrap();
        let mut prevcount : i64 = 1;

        let mut total_count = 0;

        while let Some(c) = iterator.next() {
            if prevchar == c {
                prevcount += 1;
                continue;
            }
            total_count += (((prevcount)*(prevcount + 1))/2)% modulo;
            // dbg!(prevcount);
            prevchar = c;
            prevcount = 1;
        }
        total_count += (((prevcount)*(prevcount + 1))/2) % modulo;
        return (total_count % modulo) as i32;
    }
}

fn main() {
    dbg!(Solution::count_homogenous("abbcccaa".to_owned()));
}
struct Solution;